croquemort/webhook.py

```diff
@@ -0,0 +1,74 @@
+import logbook
+import requests
+
+from nameko.dependency_providers import Config
+from nameko.events import event_handler
+from nameko.utils.retry import retry
+
+from .logger import LoggingDependency
+from .storages import RedisStorage
+from .tools import is_url
+
+log = logbook.debug
+
+# See http://docs.python-requests.org/en/latest/user/advanced/#timeouts
+# We are waiting 3 sec for the connexion and 9 for the response.
+TIMEOUT = (3.1, 9.1)
+
+DELAY = 10
+RETRY = 5
+# increase the retry delay by this factor at each try
+BACKOFF = 2
+
+
+class WebhookUnreachableException(Exception):
+
+    def __init__(self, message, original_exception=None):
+        super().__init__(message)
+        self.original_exception = original_exception
+
+
+class WebhookService(object):
+    name = 'webhook_dispatcher'
+    storage = RedisStorage()
+    logger = LoggingDependency()
+    config = Config()
+
+    def _send(self, url, metadata):
+        """POST metadata to url"""
+        try:
+            response = requests.post(url, json={'data': metadata},
+                                     timeout=TIMEOUT)
+        except (requests.Timeout, requests.RequestException) as e:
+            raise WebhookUnreachableException('Unreachable',
+                                              original_exception=e)
+        if response.status_code < 200 or response.status_code >= 400:
+            raise WebhookUnreachableException('Unreachable')
+        log('Successfully called webhook {url}'.format(url=url))
+
+    @event_handler('url_crawler', 'url_crawled')
+    def send_response(self, metadata):
+        """Call a webhook with checked url results"""
+        url = metadata.get('url')
+        callback_urls = self.storage.get_webhooks_for_url(url)
+        if not callback_urls or not len(callback_urls):
+            return
+        for callback_url in callback_urls:
+            log(('Calling webhook url {callback_url} for checked url {url}'
+                 .format(callback_url=callback_url, url=url)))
+            if not is_url(callback_url):
```

Comment:
Don't we want to check this when the callback URL is submitted instead of when we're about to submit the callback? Is that a feedback we want to give to the submitter?